File: HideAndSeek/src/main/java/com/asus/hideandseek/Seeking.java
package com.asus.hideandseek;

import android.os.Handler;

import java.util.List;

import com.asus.robotframework.API.RobotFace;
import com.asus.robotframework.API.SpeakConfig;
import com.asus.robotframework.API.Utility;
import com.asus.robotframework.API.VisionConfig;
import com.asus.robotframework.API.WheelLights;
import com.asus.robotframework.API.results.DetectFaceResult;
import com.asus.robotframework.API.results.DetectPersonResult;
import com.asus.robotframework.API.results.RecognizePersonResult;

import static com.asus.hideandseek.HideAndSeek.robotAPI;
import static com.asus.hideandseek.HideAndSeek.seeking;
import static com.asus.hideandseek.HideAndSeek.speaking;
import static com.asus.hideandseek.HideAndSeek.statusText;
import static com.asus.hideandseek.Seeking.SeekingState.ASKING_TO_PLAY;
import static com.asus.hideandseek.Seeking.SeekingState.SEEKING;
import static com.asus.robotframework.API.SpeakConfig.MODE_DEFAULT;

public class Seeking {
    public static int DEFAULT_COUNTDOWN_TIME = 5;
    public static int DEFAULT_COUNTDOWN_INTERVAL_MS = 2000;
    public enum SeekingState {
        NOT_STARTED, SEARCHING_FOR_TARGET, ASKING_TO_PLAY, COUNTDOWN, POSITIONING, SEEKING
    }

    public SeekingState state = SeekingState.NOT_STARTED;
    private int denialCount = 0;

    private boolean debugMode = true;
    private int countdownSeconds = 0;
    private VisionConfig.PersonDetectConfig detectPersonConfig = new VisionConfig.PersonDetectConfig();
    private VisionConfig.FaceDetectConfig detectFaceConfig = new VisionConfig.FaceDetectConfig();
    private VisionConfig.PersonRecognizeConfig personRecognizeConfig = new VisionConfig.PersonRecognizeConfig();
    Handler delayHandler = new Handler();
    Runnable repeatCountdown = new Runnable() {
        @Override
        public void run() {
            countdown();
        }
    };

    Seeking() {
        // Set-up the vision detections parameters
        detectPersonConfig.interval = 3;
        detectFaceConfig.interval = 3;
        detectPersonConfig.enableDebugPreview = debugMode;
        detectFaceConfig.enableDebugPreview = debugMode;
        detectPersonConfig.enableDetectHead = true;
        personRecognizeConfig.interval = 2;
        personRecognizeConfig.enableCandidateObj = false;
        personRecognizeConfig.enableDebugPreview = debugMode;
        personRecognizeConfig.enableDetectHead = true;
    }

    public void handlePersonDetection(List<RecognizePersonResult> people) {
        switch (state) {
            case SEARCHING_FOR_TARGET:
                if (people.size() > 1) {
                    HideAndSeek.speaking.say("There are too many of you!");
                    robotAPI.robot.setExpression(RobotFace.SHOCKED);
                    blinkAllLights(0x00ff0000, 5, 20);
                } else if (people.size() == 1) {
                    robotAPI.robot.setExpression(RobotFace.CONFIDENT);
                    HideAndSeek.speaking.askForPlay();
                }
                break;

            case SEEKING:
                if (people.size() == 1) {
                    String uuid = people.get(0).getUuid();
                    String name = robotAPI.contacts.family.getName(uuid);
                    boolean isValidName = name != null && uuid.charAt(0) != '-';
                    String message = (isValidName ? "Hey " + name + "!" : "Hey you there!") + " I found you!";
                    HideAndSeek.robotAPI.robot.speak(message);
                    robotAPI.robot.setExpression(RobotFace.HAPPY);
                    blinkAllLights(0x001111ff, 5, 50);
                    stop();
                }
                state = SeekingState.NOT_STARTED;
                break;

            default:
                break;
        }
    }

    public void handleFaceDetection(List<DetectFaceResult> faces) {

    }

    public void startLookingForTarget() {
        switchToPersonDetection();
        state = SeekingState.SEARCHING_FOR_TARGET;
    }

    public void handleUserAnswer(Boolean decision) {
        if (decision) {
            blinkAllLights(0x0011ff00, 2, 10);
            if (denialCount < 2) {
//                HideAndSeek.robotAPI.utility.playEmotionalAction(RobotFace.ACTIVE, Utility.PlayAction.Head_up_1);
                HideAndSeek.robotAPI.robot.speak("I'm counting until five.");
                robotAPI.robot.setExpression(RobotFace.TIRED);
                statusText.setText(seeking.state.toString());
                startCountdown(DEFAULT_COUNTDOWN_TIME);
            } else {
//                robotAPI.utility.playEmotionalAction(RobotFace.SERIOUS, Utility.PlayAction.Head_up_1);
                HideAndSeek.robotAPI.robot.speak("Well good for you, but I don't want to play with you anymore now!");
                robotAPI.robot.setExpression(RobotFace.PROUD);
                state = SeekingState.NOT_STARTED;
            }
        } else {
//            robotAPI.utility.playEmotionalAction(RobotFace.INNOCENT, Utility.PlayAction.Head_down_1);
            if (denialCount == 0) {
                HideAndSeek.robotAPI.robot.speak("Huh. Ok then.");
            }
            else {
                HideAndSeek.robotAPI.robot.speak("Again ? Ok then.");
            }
            blinkAllLights(0x00ff8800, 2, 10);
            denialCount++;
            state = SeekingState.NOT_STARTED;
        }
    }

    public void handleApology() {
//        robotAPI.robot.setExpression(RobotFace.ACTIVE)
        HideAndSeek.robotAPI.robot.speak("It's okay!");
        state = SeekingState.NOT_STARTED;
        blinkAllLights(0x000055ff, 3, 50);

        denialCount = 0;
    }

    public void startCountdown(int seconds) {
        state = SeekingState.COUNTDOWN;
        countdownSeconds = 0;
        delayHandler.postDelayed(repeatCountdown, 2000);
    }

    private void countdown() {
        countdownSeconds++;
        HideAndSeek.robotAPI.robot.speak(Integer.toString(countdownSeconds));

        if (countdownSeconds >= DEFAULT_COUNTDOWN_TIME) {
            state = SeekingState.POSITIONING;
            countdownSeconds = 0;
            Point startPoint = new Point(11, 11);
            HideAndSeek.navigation.startSearchingRoom(startPoint);
//            state = SEEKING;
        } else {
            delayHandler.postDelayed(repeatCountdown, 2000);
        }
    }

    private void blinkAllLights(int color, int cycles, int speed) {
        HideAndSeek.robotAPI.wheelLights.setColor(WheelLights.Lights.SYNC_BOTH, 0xff, color);
        HideAndSeek.robotAPI.wheelLights.startBlinking(WheelLights.Lights.SYNC_BOTH, 0xff, speed, speed, cycles);
    }

    public void switchToPersonDetection() {
        HideAndSeek.robotAPI.vision.cancelDetectFace();
        HideAndSeek.robotAPI.vision.requestRecognizePerson(personRecognizeConfig);
    }

    public void switchToFaceDetection() {
        HideAndSeek.robotAPI.vision.cancelRecognizePerson();
        HideAndSeek.robotAPI.vision.requestDetectFace(detectFaceConfig);
    }

    public void stop() {
        HideAndSeek.robotAPI.vision.cancelRecognizePerson();
        HideAndSeek.robotAPI.vision.cancelDetectFace();
        HideAndSeek.navigation.stopAllMovement();
        state = SeekingState.NOT_STARTED;
    }
}

File: app/src/main/java/com/example/zenbohideandseek/Seeking.kt
package com.example.zenbohideandseek

import android.os.Handler
import com.asus.robotframework.API.RobotAPI
import com.asus.robotframework.API.SpeakConfig
import com.asus.robotframework.API.VisionConfig
import com.asus.robotframework.API.WheelLights
import com.asus.robotframework.API.results.DetectPersonResult

class Seeking(val robotAPI: RobotAPI) {
    val DOMAIN = "85848A87233A4268AE9E6DBE15A73273"

    enum class SeekingState {
        NOT_STARTED, SEARCHING_FOR_TARGET, ASKING_TO_PLAY, COUNTDOWN, SEEKING
    }
    var state: SeekingState = SeekingState.NOT_STARTED
    var denialCount = 0
    var answerWaitIterations = 0
    var countdownSeconds = 5

    val debugMode = true
    val detectPersonConfig = VisionConfig.PersonDetectConfig()
    val detectFaceConfig = VisionConfig.FaceDetectConfig()

    init {
        // Set-up the vision detections parameters
        detectPersonConfig.interval = 3
        detectFaceConfig.interval = 3
        detectPersonConfig.enableDebugPreview = debugMode
        detectFaceConfig.enableDebugPreview = debugMode
        detectPersonConfig.enableDetectHead = true
    }

    fun handlePersonDetection(people: List<DetectPersonResult>) {
        when (state) {
            SeekingState.SEARCHING_FOR_TARGET -> {
                if (people.size > 1) {
                    robotAPI.robot.speak("There are too many of you!")
                    blinkAllLights(0x00ff0000, 5, 20)
                } else if (people.size == 1) {

                    val speakingConfig = SpeakConfig()
                    speakingConfig.domain(DOMAIN)
                    speakingConfig.timeout(10f)
                    speakingConfig.listenCurrentDomain(true)

                    state = SeekingState.ASKING_TO_PLAY
                    robotAPI.robot.stopSpeakAndListen()
                    robotAPI.robot.speakAndListen("Do you want to play hide and seek with me?",
                            speakingConfig)
                }
            }
            SeekingState.ASKING_TO_PLAY -> {
                if (people.isEmpty()) {
                    robotAPI.robot.speak("Where did you go?")
                    robotAPI.robot.speak("Don't you know it's a bit rude to walk away from the conversation?")
                    state = SeekingState.ASKING_TO_PLAY
                }
                answerWaitIterations++
//                if (answerWaitIterations > 20) {
//                    answerWaitIterations = 0
//
//                    robotAPI.robot.speak("Hey, I asked you something!")
//                }
            }
            SeekingState.SEEKING -> {
                if (people.size == 1) {
                    robotAPI.robot.speak("Haha, I found you!")
                    blinkAllLights(0x001111ff, 5, 50)
                }
                state = SeekingState.NOT_STARTED
            }
        }
    }

    fun resetState() {
        state = SeekingState.NOT_STARTED
    }

    fun startLookingForTarget() {
        println("Zenbo will now look for a person and ask him.")
//        robotAPI.robot.setExpression(RobotFace.EXPECTING)
        state = SeekingState.SEARCHING_FOR_TARGET
    }

    fun handleUserAnswer(decision: Boolean) {
        if (decision) {
            blinkAllLights(0x0011ff00, 2, 10)
            if (denialCount < 2) {
//                robotAPI.utility.playEmotionalAction(RobotFace.ACTIVE, Utility.PlayAction.Head_up_1)
                robotAPI.robot.speak("Okay! Hide now, I'm counting until five.")
                startCountdown(5)
            } else {
//                robotAPI.utility.playEmotionalAction(RobotFace.SERIOUS, Utility.PlayAction.Head_up_1)
                robotAPI.robot.speak("Well good for you, but I don't want to play with you anymore now!")
                state = SeekingState.NOT_STARTED
            }
        } else {
//            robotAPI.utility.playEmotionalAction(RobotFace.INNOCENT, Utility.PlayAction.Head_down_1)
            robotAPI.robot.speak(if (denialCount == 0) "Huh. Ok then." else "Again? Ok then.")
            blinkAllLights(0x00ff8800, 2, 10)
            denialCount++
        }
    }

    fun handleApology() {
//        robotAPI.robot.setExpression(RobotFace.ACTIVE)
        robotAPI.robot.speak("It's okay!")

        blinkAllLights(0x000055ff, 3, 50)

        denialCount = 0
    }

    fun startCountdown(seconds: Int) {
        state = SeekingState.COUNTDOWN
        countdownSeconds = seconds
        announceCountdown()
    }

    private fun announceCountdown() {
        robotAPI.robot.speak(countdownSeconds.toString())
        blinkAllLights(0x00005500, 1, 50)

        if (countdownSeconds > 0) {
            Handler().postDelayed({ announceCountdown() }, 2000)
        } else {
            robotAPI.robot.speak("I'm coming!")
            state = SeekingState.SEEKING
        }

        countdownSeconds--
    }

    private fun blinkAllLights(color: Int, cycles: Int, speed: Int) {
        robotAPI.wheelLights.setColor(WheelLights.Lights.SYNC_BOTH, 0xfffffff, color)
        robotAPI.wheelLights.startBlinking(WheelLights.Lights.SYNC_BOTH, 0xfffffff, speed, speed, cycles)
    }

    fun switchToPersonDetection() {
        robotAPI.vision.cancelDetectFace()
        robotAPI.vision.requestDetectPerson(detectPersonConfig)
    }

    fun switchToFaceDetection() {
        robotAPI.vision.cancelDetectPerson()
        robotAPI.vision.requestDetectFace(detectFaceConfig)
    }

    fun stop() {
        robotAPI.vision.cancelDetectPerson()
        robotAPI.vision.cancelDetectFace()
        state = SeekingState.NOT_STARTED
    }
}